### ISSUE TITLE
Docker for Mac: vhost-proxy and dns system services are not running

### ISSUE THREAD

**Issue by lmakarov** [docksal/docksal#394]
In the past week, I noticed `docksal-vhost-proxy` and `docksal-dns` services were dead on several occasions. I first thought they did not restart properly, however, by restart  Docker for Mac I was not able to reproduce this.

Looking at the container logs, this looks to be a shutdown request from the host:

```
...
2017-11-18 22:38:26,109 WARN received SIGTERM indicating exit request
2017-11-18 22:38:26,185 INFO waiting for dnsmasq, docker-gen to die
2017/11/18 22:38:26 Received event die for container d8cc9ddf518f
2017/11/18 22:38:26 Error retrieving docker server info: cannot connect to Docker endpoint
2017/11/18 22:38:26 Error listing containers: cannot connect to Docker endpoint
2017/11/18 22:38:26 Received signal: terminated
2017/11/18 22:38:26 Received signal: terminated
2017-11-18 22:38:26,190 INFO stopped: docker-gen (exit status 0)
dnsmasq: exiting on receipt of SIGTERM
2017-11-18 22:38:26,201 INFO stopped: dnsmasq (exit status 0)
```

Inspecting the containers shows this:

```
...
        "State": {
            "Status": "exited",
            "Running": false,
            "Paused": false,
            "Restarting": false,
            "OOMKilled": false,
            "Dead": false,
            "Pid": 0,
            "ExitCode": 128,
            "Error": "driver failed programming external connectivity on endpoint docksal-dns (3d57a839f096aff6dcd9d83f30430a03f904abcb3e4e6aaa2e4ab9ff4d1abf73): Error starting userland proxy: Bind for 192.168.64.100:53: unexpected error (Failure EADDRNOTAVAIL)",
            "StartedAt": "2017-11-18T07:52:16.538254853Z",
            "FinishedAt": "2017-11-18T22:38:26.344627967Z"
        },
...
```

**Comment by achekulaev** [comment]
I've created a commit ^ with functionality that was waiting to be added long time ago. 
For some commands (like start, reset etc) Check if Docksal system services are running. Restart them if they are not.

**Comment by achekulaev** [comment]
So looks like there is definitely some issue but it's hard to tell what is happening exactly just yet.

I started Docker for Mac today and ran a fin command. Thanks to the commit above fin has detected that some services are down and ran `system reset` automatically. However it did not really succeed.

<img width="844" alt="demo -bash 1 2017-12-01 13-59-55" src="https://user-images.githubusercontent.com/402007/33498606-fd08e500-d69f-11e7-8169-b5adab393a85.png">


```
$ fin ps
 ALERT:  Some Docksal services were not running
         Restarting Docksal system services...
Resetting network settings...
Password:
Password:
Resetting Docksal services...
 * proxy
docker: Error response from daemon: driver failed programming external connectivity on 
endpoint docksal-vhost-proxy (f17a0c73dc2f559f22a4ede8c275cade114c02cbbcd273d1b732a0861dcc37a8): 
Error starting userland proxy: Bind for 192.168.64.100:443: unexpected error (Failure EADDRNOTAVAIL).
 * dns and resolver for .docksal domain
docker: Error response from daemon: driver failed programming external connectivity on 
endpoint docksal-dns (cfcb996fa92e060ebf8b59250b8f7e1a110189cad86ecdda378468d0571b6201): 
Error starting userland proxy: Bind for 192.168.64.100:53: unexpected error (Failure EADDRNOTAVAIL).
Enabling automatic *.docksal DNS resolver...
Clearing DNS cache...
 * ssh-agent
Identity added: id_rsa (id_rsa)
Identity added: id_dsa (id_dsa)
Name   Command   State   Ports
------------------------------
```

You can notice those Address not available errors.

I checked running containers.

```
**$ fin d ps**
CONTAINER ID        IMAGE                   COMMAND               CREATED             STATUS              PORTS               NAMES
aa1042c2cd2f        docksal/ssh-agent:1.0   "/run.sh ssh-agent"   12 seconds ago      Up 10 seconds                           docksal-ssh-agent
```

Indeed. Only SSH Agent is running. So I tried again manually, and it worked fine. This time. 

```
$ fin reset system
Resetting network settings...
Resetting Docksal services...
 * proxy
 * dns and resolver for .docksal domain
Enabling automatic *.docksal DNS resolver...
Clearing DNS cache...
 * ssh-agent
Identity added: id_rsa (id_rsa)
Identity added: id_dsa (id_dsa)
```

So I wonder what could be the reason why it can't bind to that address.

**Comment by lmakarov** [comment]
@achekulaev this could be due to both of us having VirtualBox and Docker for Mac on a single machine.

```
$ ifconfig | grep -B5 -A5 192.168.64.
lo0: flags=8049<UP,LOOPBACK,RUNNING,MULTICAST> mtu 16384
	options=1203<RXCSUM,TXCSUM,TXSTATUS,SW_TIMESTAMP>
	inet 127.0.0.1 netmask 0xff000000 
	inet6 ::1 prefixlen 128 
	inet6 fe80::1%lo0 prefixlen 64 scopeid 0x1 
	inet 192.168.64.100 netmask 0xffffff00 
	inet 192.168.64.1 netmask 0xffffff00 
	nd6 options=201<PERFORMNUD,DAD>
gif0: flags=8010<POINTOPOINT,MULTICAST> mtu 1280
stf0: flags=0<> mtu 1280
en0: flags=8863<UP,BROADCAST,SMART,RUNNING,SIMPLEX,MULTICAST> mtu 1500
	ether 80:e6:50:25:c1:a2 
--
--
utun0: flags=8051<UP,POINTOPOINT,RUNNING,MULTICAST> mtu 2000
	inet6 fe80::5510:d0ee:ba5f:b59d%utun0 prefixlen 64 scopeid 0xa 
	nd6 options=201<PERFORMNUD,DAD>
vboxnet0: flags=8843<UP,BROADCAST,RUNNING,SIMPLEX,MULTICAST> mtu 1500
	ether 0a:00:27:00:00:00 
	inet 192.168.64.1 netmask 0xffffff00 broadcast 192.168.64.255
utun1: flags=8051<UP,POINTOPOINT,RUNNING,MULTICAST> mtu 1290
	inet 10.35.153.254 --> 10.35.153.254 netmask 0xffffffff
```

When we assign `192.164.100.1` and `192.168.64.100` to `lo0` we do not kill `vboxnet0`, which still utilizes the same subnet.

We should probably kill the vbox interface serving Docksal VBox VM in `fin reset network`, when switching to Docker for Mac mode.

**Comment by jrglasgow** [comment]
any progress on this, I have been having a similar problem for the last 2 days

**Comment by achekulaev** [comment]
fin should restart them automatically now so even if it happens it should not be a huge problem. However I should note that I use Docker for Mac daily and haven'e experienced the issue.

**Comment by lmakarov** [comment]
I ran into this today. I think it happened after I was switching to VirtualBox and then back to Docker for Mac. 

![image](https://user-images.githubusercontent.com/1205005/34173343-01fdf678-e4aa-11e7-9108-ab488e34dc76.png)

After hitting `y` fin restarted them and everything is back now.

We should try what I suggested in https://github.com/docksal/docksal/issues/394#issuecomment-348590339

@jrglasgow do/did you also use Docker with VirtualBox and Docker for Mac on the same machine?

**Comment by achekulaev** [comment]
@lmakarov we will hit it regularly with Docker for Windows. It runs as user so it will not be able to auto-restart services that bind to ports 80 and 443.

**Comment by lmakarov** [comment]
I just realized (after running into this issues again), that `ifconfig lo0 alias` does not persist after host reboot:

```
$ ifconfig lo0
lo0: flags=8049<UP,LOOPBACK,RUNNING,MULTICAST> mtu 16384
	options=1203<RXCSUM,TXCSUM,TXSTATUS,SW_TIMESTAMP>
	inet 127.0.0.1 netmask 0xff000000 
	inet6 ::1 prefixlen 128 
	inet6 fe80::1%lo0 prefixlen 64 scopeid 0x1 
	nd6 options=201<PERFORMNUD,DAD>
```

Docksal's canonical IPs (`192.168.64.1` and `192.165.64.100`) will always be lost after host reboot.

Apparently, there is no standard way for defining persistent aliases in macOS. Defining a [login hook](https://support.apple.com/en-us/HT2420) looks to be an option.

**Comment by achekulaev** [comment]
1. A workaround has been added into fin that will start system services if fin detects that they are not running
2. The issue itself has been fixed in 17.12 which now asks for admin password to be able to bind to ports <1024.
